### ISSUE TITLE
Skyrim: Special Edition + ENB crash under certain conditions

### ISSUE THREAD

**Issue by ErikSteiner** [doitsujin/dxvk#2033]
Unlike [#865](https://github.com/doitsujin/dxvk/issues/865) it is possible to use ENB v0.452. It is only required to set `UseEffect=false` in `enbseries.ini`. During the Bethesda Into one is able to press the key combination in order to show the ENB GUI. In the GUI you can change `UseEffect` to `true`, however this is only possible if one is not in the following places:
- main menu
- map
- loading screen
- skills menu
- hitting the ESC key

The following places are not affected by a game crash if ENB is enabled:
- worldspace
- inventory
- magic (inventory?)

Creating a trace was a bit complicated, as the ENB uses its own `d3d11.dll`. This resulted in an instant crash at game startup, so I used the `ProxyLibrary` feature of the ENB to load the trace-dll from `apitrace-10.0-win64/lib/wrappers`. The traces have a size of 360 MB and 2,5 GB.

### Software information
- Skyrim Special Edition
- ENB v0.452 (d3d11.dll, d3dcompiler_46e.dll)
- [Simply Cathedral 1.0](https://www.nexusmods.com/skyrimspecialedition/mods/31367)

### System information
- GPU: NVIDIA GTX 1080 Ti
- Driver: 460.67
- Wine version: Proton-6.1-GE-1
- DXVK version: 1.8.1

### Apitrace file(s)
- [SkyrimSE.trace_main-menu](https://mega.nz/file/WM4mBbqQ#uc4_cbvCeHdS7HaQrWYkGpfFUuEU_BpVdKI-flL3Y18)

### Log files
[SkyrimSE_d3d11.log](https://mega.nz/file/PZxAxTKR#-ADw2y1DFwiuvc2A1VV5tqU3jIkQR9Z3U_Ctc8iTVN0)
[SkyrimSE_dxgi.log](https://github.com/doitsujin/dxvk/files/6346602/SkyrimSE_dxgi.log)

**Comment by doitsujin** [comment]
Do we know what the issue is exactly? I really don't want to waste even more time on the horrendous shitshow that is ENB, it does some really weird shit with shaders that we just cannot support, and it's a miracle that it doesn't just crash and burn on native drivers.

**Comment by ErikSteiner** [comment]
The author of ENB recently released a Linux version of 0.452. Basically he disabled an effect in the dll itself and asked a subject if the CTD in main menu is fixed. If not he disabled another thing and so on. All effects that resulted in a crash on Skyrim SE on Linux with DXVK. I guess he said something of Skylighting effects that are absent.

It is noticeable that in the Linux version only the d3d11.dll is required, whereas the Windows version also includes a d3dcompiler_46e.dll. I don't think I need to mention that it sounds completely crazy that disabling sky effects will fix the problem. I mean, why do I need them in the main menu or in other menus? However, the mentioned effect is only a guess. He does not play with open cards in this regard and is not otherwise willing to address the issue.

Is the trace file usable? I have another 2,5 GB trace, where I activated the Skill menu while `UseEffect` was true. It resulted in a crash as I pointed out, because it is one of these places.

If the trace file is usable, I could create another trace with the Linux version that don't cause a CTD in main menu if `UseEffect` is true?

**Comment by doitsujin** [comment]
The trace works, yes, and it shows one problem that can definitely cause crashes, gonna fix that and hope that it works.

**Comment by doitsujin** [comment]
Does https://github.com/doitsujin/dxvk/tree/range-checks fix the issue for you?

**Comment by ErikSteiner** [comment]
I am so happy to hear from you. I tested it out,

1. first by replacing the dll's in
`~/.steam/root/compatibilitytools.d/Proton-6.1-GE-1/dist/lib/wine/dxvk`
`~/.steam/root/compatibilitytools.d/Proton-6.1-GE-1/dist/lib64/wine/dxvk`
with the dlls of `./dxvk-range-checks/x32` and `./dxvk-range-checks/x64` and appropriate permissions 555.
2. I launched the game with `UseEffect=false` and hit the key combination for the ENB GUI during the intro and entered main menu. In the main menu I activated the effects in ENB GUI and the game crashed.

So far, nothing has changed in the fact that the game crashes in the aforementioned location. Furthermore, I also tested to open the map ingame, but again the same result. I uploaded a new trace file and logs. Thank you for your effort.

### Apitrace file(s)
[SkyrimSE.trace](https://mega.nz/file/XZ5GGaiB#W78bFcp0WKkRtEY9WICItk6jQI4K6wcEPDUL_N_Z45Y)

### Log files
- d3d11.log: [SkyrimSE_d3d11.log](https://mega.nz/file/PZomlagQ#98xSG5KYz9K6gXl64C64FohfvjVS79Y60TZOfsI8b4M)
- dxgi.log: [SkyrimSE_dxgi.log](https://github.com/doitsujin/dxvk/files/6353648/SkyrimSE_dxgi.log)

**Comment by doitsujin** [comment]
Yeah, no idea what to do then. The trace you attached replays perfectly fine on my end, not really sure how to debug this further.

**Comment by ErikSteiner** [comment]
This is very disappointing. I fired all logs onto that thing. Maybe there is something related to the issue? So, I load the game with UseEffect=false, activate the ENB GUI during the intro and activate the Effect when in main menu.

### Apitrace file(s)
[SkyrimSE.trace](https://mega.nz/file/qQoxAQ5K#aGHzKxtLhmxJVfboeq6XOwRetuRMp0uEMtc-mf2KjAw)
### Log files
[steam-489830.log](https://github.com/doitsujin/dxvk/files/6360630/steam-489830.log)
[user_settings.py](https://github.com/doitsujin/dxvk/files/6360648/user_settings.txt) for proton

**Comment by jellybeanmonster** [comment]
This issue applies to me somewhat except i cant even go to exterior places in skyrim without crashing. Im only able to be in interiors when i enable the UseEffect option on enb. All the menus in game give instant crash.
More information:
GPU: RTX 3080 10gb
CPU: AMD Ryzen 9 5900x
DXVK version: 1.8.1
[SkyrimSE_d3d11.log ](https://drive.google.com/file/d/1SKf7g8oppKGYkVbwuPo83ccwoZNgHcet/view?usp=sharing)
[SkyrimSE_dxgi.log](https://drive.google.com/file/d/1u41DNVM8ZsN4Csanqf0538PutVcRVwzI/view?usp=sharing)

**Comment by ErikSteiner** [comment]
Are you able to test [this](https://github.com/doitsujin/dxvk/issues/2033#issuecomment-824020074) build? Moreover, could you create a trace?

1. To create a trace, I downloaded the apitrace [file](https://github.com/apitrace/apitrace/releases/download/10.0/apitrace-10.0-win64.7z)
2. In `../apitrace-10.0-win64/lib/wrappers` I renamed the `d3d11.dll` to `d3d11_trace.dll` and copied `d3d11_trace.dll`, `dxgi.dll` and `dxgitrace.dll` to `../common/Skyrim Special Edition`
3. In enblocal.ini I set the proxy to
```
[PROXY]
EnableProxyLibrary=true
InitProxyFunctions=true
ProxyLibrary=d3d11_trace.dll
```

**Comment by jellybeanmonster** [comment]
Ok im about to create a trace . was i supposed to try that build before i create a trace?

**Comment by ErikSteiner** [comment]
It would be nice to have both situations. One trace with the current release and your configuration and one with the new build of DXVK.

**Comment by jellybeanmonster** [comment]
what do i copy and paste from the dxvk range checks ? thats what i got from the link.

**Comment by ErikSteiner** [comment]
Please feel free to either replace the DXVK dll's like I [did](https://github.com/doitsujin/dxvk/issues/2033#issuecomment-824332657), or by installing the package with `./setup_dxvk.sh install --with-d3d10 `. For the latter, please make sure that you export the path to the prefix, before executing the sh. Otherwise .wine is the install location. So do `export WINEPREFIX="/home/user/.steam/root/steamapps/compatdata/489830/pfx"` or similar.

**Comment by doitsujin** [comment]
I mean, creating traces with different DXVK builds is probably not going to change anything since apitrace records D3D11 commands, which are *usually* unaffected by what DXVK is doing and that is the case here.

**Comment by ErikSteiner** [comment]
So that means, that we have a CTD that is not able to be tracked by a trace? Or is it possible to trace it in another mean?

**Comment by doitsujin** [comment]
you'd need to trace on Windows.

**Comment by ErikSteiner** [comment]
Would you be so kind, to give an advice, how to do it. I have a second boot, but that is out of my knowledge.

**Comment by doitsujin** [comment]
Same as on Linux I guess? You'll have the same problems with ENB though which makes the whole thing a real pain in the arse to work with.

**Comment by jellybeanmonster** [comment]
Coming back with bad and somewhat good news about using the trace files. Ok so the "new build" i downloaded wouldnt even let me start the game so i had to go back and get my original d3d11.ll file from my enb to even start the game. Once i was able to get into the game my framerate literally stayed on 1-2 fps. When loaded into the interior space it still was going at 1-2 fps. i was able to open the enb menu and turn on use effect without issue. The somewhat "good" news is when i went outside to the exterior  i didnt crash at all but i still was going  at 1-2 fps. Menus didnt crash at all either. Also is the trace files supposed to spit out some logs?

**Comment by ErikSteiner** [comment]
I am not sure if I did it right. The trace is created in Windows by the following procedure. There is no DXVK involved:

1. To create a trace, I downloaded the apitrace file
2. In ../apitrace-10.0-win64/lib/wrappers I renamed the d3d11.dll to d3d11_trace.dll and copied d3d11_trace.dll, dxgi.dll and dxgitrace.dll to ../common/Skyrim Special Edition
3. In enblocal.ini I set the proxy to
```
[PROXY]
EnableProxyLibrary=true
InitProxyFunctions=true
ProxyLibrary=d3d11_trace.dll
```
### Apitrace file(s)
[SkyrimSE.trace](https://mega.nz/file/icxVQQ5C#d00wZLv4xa0L9DklnzyVDupLGqngRlm_m0FhOFObtG4)

**Comment by jellybeanmonster** [comment]
I got 3 trace files on my desktop that i didnt notice until now. It is 8gb for some reason. im uploading it to MEGA right now incase its needed her i guess.

**Comment by jellybeanmonster** [comment]
Those are the trace files i got.
[SkyrimSE.trace](https://mega.nz/file/1kBXSAqT#KYut6xrtsfREFt286ajqBCg1Qbx4QOnN8wZLhROE2B0)

**Comment by ErikSteiner** [comment]
I updated my [post](https://github.com/doitsujin/dxvk/issues/2033#issuecomment-825980882). If you have access to a windows machine, feel free to reproduce my steps. The trace file will be written to you Desktop. @doitsujin I hope that there are now better information in these tracing files. For the DXVK variant I used you provided build.

**Comment by medinaKrom1** [comment]
Hi and good evening @doitsujin and @ErikSteiner,
I'm new here and on github too and really don't want brake any Issue discussion rules, _so sorry for my question_.
Discovered the great work result of **dxvk** a bit and found Erik's Issue.

**So the question is: Can i use Skyrim SE with the DLL's of dxvk together with the current version of ENB files?**
Or in other words: How i have to combine the ENB files with the files from dxvk V1.8

From ENB & game side i have the files from Skyrim modlist Elysium from Wabbajack:
d3d11.dll - File Version: 0.454
d3dcompiler_46e.dll
d3dcompiler_47e.dll
d3dx9_42.dll
dxgi.dll - For Reshade File VErsion: 4.9.1.1003
skse64_1_5_97.dll
(If i'm using dxgi.dll and d3d11.dll from dxvk, the ENB is not working.)

Hardware: AMD Ryzen 5800X and RX 6800XT graphics card.
Maybe I'm running against a wall and the use case is generally not possible.

Thank you in advance for any tips, where I hopefully can continue.

**Comment by chumik6000** [comment]
> Does https://github.com/doitsujin/dxvk/tree/range-checks fix the issue for you? (build [here](https://github.com/doitsujin/dxvk/actions/runs/770515241))

The problem is gone with this fix for me (latest enb).

**Comment by ErikSteiner** [comment]
@chumik6000: Could you add information to your environment? Do you run it under Linux (Kernel, Proton/Wine version, winetricks etc.) or under Windows (setup).

**Comment by chumik6000** [comment]
Windows 10, AMD RX 6900xt, latest AMD drivers.

**Comment by Chaddabox** [comment]
Wondering if anyone has experienced this??:

-im in Skyrim, press shift + enter to open ENB. 
-i had a cursor problem and typed “tm” into commands
-pressed shift+enter to exit enb
-now i am unable to use any ingame menus (tab, esc) i can press m for maps but it doesnt show anything on it
-i also dont have a compass/ui
i have tried typing “tm” to reverse it and the only thing that changes is i dont have a cursor on enb

**Comment by Blisto91** [comment]
@ErikSteiner Friendly ping. Are you still having this issue with latest dxvk, ENB & dirvers?

**Comment by hotburger** [comment]
I have been playing with it for a while and it seems stable to me. ENB 0.488, mesa 23.1, and Proton 8.0-2 (DXVK 2.1-4).
Winetricks d3dcompiler_47, d3dcompiler_43, and d3dx11_43.

mesa GPL was causing error messages so I disabled it with `RADV_DEBUG=nogpl`

**Comment by Blisto91** [comment]
What error messages

**Comment by hotburger** [comment]
```
Assertion failed!
Program Z:\mnt\ssd\Steam\steamapps\common\Skyrim Special Edition\SkyrimSE.exe
File: ../src-wine/dlls/winevulkan/loader_thunks.c
Line: 2703
Expression: "!status"
```
7 of these identical popups come up on game launch with GPL enabled. Sorry, I'm not sure how to get a more detailed error message. I can actually still play the game with these popups up and it doesn't crash (though i haven't tested it very much).

**Comment by ErikSteiner** [comment]
> @ErikSteiner Friendly ping. Are you still having this issue with latest dxvk, ENB & drivers?
I was without a PC for a long time, so I'm just getting back to you. With the following parameters, Skyrim now works under Linux and ENB:

**Software information**
    Skyrim Special Edition
    ENB v0.493 (d3d11.dll, d3dcompiler_46e.dll)
    [Simply Cathedral 1.0](https://www.nexusmods.com/skyrimspecialedition/mods/31367)

**System information**
    GPU: NVIDIA GTX 1080 Ti
    Driver: 530.41.03
    Wine version: Proton-8.3-GE
    DXVK version: 2.1
    SteamTinkerLauncher:
- winecfg: override d3d11 as native
- winetricks: d3dx11_43 d3dcompiler_43 d3dcompiler_47

I did not notice any difference between enabled and disabled dxvk_async. The ENB GUI could be opened at any time. However, I noticed that with night mode enabled in Gnome, Skyrim ends up in a black screen. Here only a restart helps, because apparently Xorg or something else crashes.

**Comment by onlyreportingissues** [comment]
RX 6600 user here & using Proton-GE.

Didn't have to do any winetricks, winecfg & dxvk.conf tweaks.

When starting the game, ENB alters your enbseries.ini file (if its an old one) and adds a new entry called "EnableTerrainBlending=true". Set the value to **_false_** and you should be good then.

**Comment by ErikSteiner** [comment]
> RX 6600 user here & using Proton-GE.
> 
> Didn't have to do any winetricks, winecfg & dxvk.conf tweaks.
> 
> When starting the game, ENB alters your enbseries.ini file (if its an old one) and adds a new entry called "EnableTerrainBlending=true". Set the value to **_false_** and you should be good then.
> 
> Also, delete the enbcache folder always whenever a change is made to the ENB files.
> 
> I recommend the performance preset of NAT.ENB - ENB PRESET v3.1.1C: https://www.nexusmods.com/skyrimspecialedition/mods/27141?tab=files
> 
> It works really well & no tweaking needed. You can disable the FPS counter in the enblocal.ini file & the letterbox in the enbseries/enbeffect.fx.ini file by setting the value of "Letterbox" to 0.

Are you sure, that you used the `WrapperVersion`?

**Comment by onlyreportingissues** [comment]
> > Didn't have to do any winetricks, winecfg & dxvk.conf tweaks.
> 
> Are you sure, that you used the `WrapperVersion`?

Yes, but maybe I didn't need to tweak because I'm using Proton-GE? Not sure. I don't know if it's connected, but no matter which ENB preset I'm using, it looks like Silent Hill (kinda foggy appearance colour-wise). Didn't try it on vanilla Skyrim though, modded the game before I tried ENB.

**Comment by fighuass** [comment]
> Yeah that's probably ENB producing broken DXBC shaders again which we then translate into broken SPIR-V.
> 
> Not sure how much I really _care_ at this point to be honest, just disable GPL as a workaround if that helps, you can do this with a DXVK config file as well by setting `dxvk.enableGraphicsPipelineLibrary = False`.

This fixed the immediate crashing and I can play in indoors environments now, but the error reappears whenever I go outside. What could that mean?

**Comment by onlyreportingissues** [comment]
This collection has some ENB mods & game fixes that might, there is also a guide for setting up ENB for Skyrim in the description: https://next.nexusmods.com/skyrimspecialedition/collections/2bhk0y

**Comment by doitsujin** [comment]
No idea if this is still an issue, new shader compiler *should* be more robust against broken code, either way this original report won't really apply anymore given how much things have changed.
